docs/source/api-config.rst
..
    # (The MIT License)
    #
    # Copyright (c) 2013-2021 Kura
    #
    # Permission is hereby granted, free of charge, to any person obtaining a
    # copy of this software and associated documentation files (the
    # 'Software'), to deal in the Software without restriction, including
    # without limitation the rights to use, copy, modify, merge, publish,
    # distribute, sublicense, and/or sell copies of the Software, and to permit
    # persons to whom the Software is furnished to do so, subject to the
    # following conditions:
    #
    # The above copyright notice and this permission notice shall be included
    # in all copies or substantial portions of the Software.
    #
    # THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS
    # OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    # MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
    # NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
    # DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
    # OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
    # USE OR OTHER DEALINGS IN THE SOFTWARE.

=======================
:mod:`blackhole.config`
=======================

.. module:: blackhole.config
    :platform: Unix
    :synopsis: Configuration structure and functionality for testing config
               validity.
.. moduleauthor:: Kura <kura@kura.gg>

Configuration structure and functionality for testing config validity.

.. autofunction:: parse_cmd_args

.. autofunction:: warn_options

.. autofunction:: config_test

.. autofunction:: _compare_uid_and_gid

.. autoclass:: Config
   :inherited-members:
   :member-order: bysource


docs/source/configuration.rst
..
    # (The MIT License)
    #
    # Copyright (c) 2013-2020 Kura
    #
    # Permission is hereby granted, free of charge, to any person obtaining a copy
    # of this software and associated documentation files (the 'Software'), to deal
    # in the Software without restriction, including without limitation the rights
    # to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    # copies of the Software, and to permit persons to whom the Software is
    # furnished to do so, subject to the following conditions:
    #
    # The above copyright notice and this permission notice shall be included in
    # all copies or substantial portions of the Software.
    #
    # THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    # IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    # FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    # AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    # SOFTWARE.

=============
Configuration
=============

.. _command-line-options:

Command line options
====================

Configuration options can be passed via the command line
as below:

-h			show this help message and exit
-v			show program's version number and exit
-c FILE		override the default configuration options
-t			perform a configuration test and exit
-d			enable debugging mode
-b			run in the background
-ls			Disable :py:obj:`ssl.OP_SINGLE_DH_USE` and :py:obj:`ssl.OP_SINGLE_ECDH_USE`.
			Reduces CPU overhead at the expense of security. Don't use this
			option unless you really need to. -- added in :ref:`2.0.13`
-q			Suppress warnings when using -ls/--less-secure, running as root or
			not using :ref:`tls_dhparams` option.


.. _configuration-options:

Configuration options
=====================

Here are all available options for the configuration file, their default values
and information on what the options actually do.

From :ref:`2.1.8` onwards the ``blackhole_config`` command will also display this
information directly from the command line.

- `listen`_
- `tls_listen`_
- `user`_
- `group`_
- `pidfile`_
- `timeout`_
- `tls_cert`_
- `tls_key`_
- `tls_dhparams`_
- `delay`_
- `mode`_
- `max_message_size`_
- `dynamic_switch`_
- `workers`_

-----

.. _listen:

listen
------

:Syntax:
    **listen** = *[address]:port [mode=MODE] [delay=DELAY]*
:Default:
    127.0.0.1:25, 127.0.0.1:587, :::25, :::587 -- 25 is the recognised SMTP
    port, 587 is the recognised SMTP Submission port. IPv6 listeners are only
    enabled if IPv6 is supported.
:Optional:
    *mode=* and *delay=* -- allows setting a response mode and delay per
    listener.
:Added:
    :ref:`2.0.8` -- introduced the new IPv6 aware syntax
    :ref:`2.1.4` -- added optional mode and delay flags

`:25` is equivalent to listening on port 25 on all IPv4 addresses and `:::25`
is equivalent to listening on port 25 on all IPv6 addresses.

Multiple addresses and ports can be listed on a single line.

::

    listen = 10.0.0.1:25, 10.0.0.2:25, :25, :::25, :587, :::587

The ``mode=`` and ``delay=`` flags allow specific ports to act in specific
ways. i.e. you could accept all mail on 10.0.0.1:25 and bounce it all on
10.0.0.2:25, as below.

::

    listen = 10.0.0.1:25 mode=accept, 10.0.0.2:25 mode=bounce

The ``mode=`` and ``delay=`` flags may also be specified together, as required.

::

    listen = 10.0.0.1:25 mode=accept delay=5, 10.0.0.2:25 mode=bounce delay=10

The flags accept the same options as :ref:`dynamic-switches`, including setting
a delay range.

-----

.. _tls_listen:

tls_listen
----------

:Syntax:
    **tls_listen** = *[address]:port [mode=MODE] [delay=DELAY]*
:Default:
    None -- 465 is the recognised SMTPS port [*]_.
:Optional:
    *mode=* and *delay=* -- allows setting a response mode and delay per
    listener.
:Added:
    :ref:`2.0.8` -- introduced the new IPv6 aware syntax
    :ref:`2.1.4` -- added optional mode and delay flags

`:465` is equivalent to listening on port 465 on all IPv4 addresses and
`:::465` is equivalent to listening on port 465 on all IPv6 addresses.

Multiple addresses and ports can be listed on a single line.

::

    tls_listen = 10.0.0.1:465, 10.0.0.2:465, :465, :::465

The ``mode=`` and ``delay=`` flags allow specific ports to act in specific
ways. i.e. you could accept all mail on 10.0.0.1:465 and bounce it all on
10.0.0.2:465, as below.

::

    tls_listen = 10.0.0.1:465 mode=accept, 10.0.0.2:465 mode=bounce

The ``mode=`` and ``delay=`` flags may also be specified together, as required.

::

    tls_listen = 10.0.0.1:465 mode=accept delay=5, 10.0.0.2:465 mode=bounce delay=10

The flags accept the same options as :ref:`dynamic-switches`, including setting
a delay range.

.. [*] Port 465 -- while originally a recognised port for SMTP over
   SSL/TLS -- is no longer advised for use. It's listed here because it's a
   well known and well used port, but also because Blackhole currently does not
   support ``STARTTLS`` over SMTP or SMTP Submission. --
   `<https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt>`_

-----

.. _user:

user
----

:Syntax:
    **user** = *user*
:Default:
    The current Linux user

Blackhole will set it's process owner to the value provided with this options.
Ports below 1024 require sudo or root privileges, this option is available so
that the process can be started, listen on privileged ports and then give up
those privileges.

::

    user = blackhole

-----

.. _group:

group
-----

:Syntax:
    **group** = *group*
:Default:
    The primary group of the current Linux user

Blackhole will set it's process group to the value provided with this options.

::

    group = blackhole

-----

.. _pidfile:

pidfile
-------

:Syntax:
    **pidfile** = */path/to/file.pid*
:Default:
    /tmp/blackhole.pid
:Added:
    :ref:`2.0.4`

Blackhole will write it's Process ID to this file, allowing you to easily track
the process and send signals to it.

::

    pidfile = /var/run/blackhole.pid

-----

.. _timeout:

timeout
-------

:Syntax:
    **timeout** = *seconds*
:Default:
    60 -- Maximum value of 180 seconds.

This is the amount of time to wait for a client to send data. Once the timeout
value has been reached with no data being sent by the client, the connection
will be terminated and a ``421 Timeout`` message will be sent to the client.

Helps mitigate DoS risks.

::

    timeout = 30

-----

.. _tls_cert:

tls_cert
--------

:Syntax:
    **tls_cert** = */path/to/certificate.pem*
:Default:
    None

The certificate file in x509 format for wrapping a connection in SSL/TLS.

::

    tls_cert = /etc/ssl/certs/blackhole.crt

-----

.. _tls_key:

tls_key
-------

:Syntax:
    **tls_key** = */path/to/private.key*
:Default:
    None

The private key of the `tls_cert`_.

::

    tls_key = /etc/ssl/private/blackhole.key

-----

.. _tls_dhparams:

tls_dhparams
------------

:Syntax:
    **tls_dhparams** = */path/to/dhparams.pem*
:Default:
    None
:Added:
    :ref:`2.0.4`


File containing Diffie Hellman ephemeral parameters for ECDH ciphers.

::

    tls_dhparams = /etc/ssl/dhparams.pem

-----

.. _delay:

delay
-----

:Syntax:
    **delay** = *seconds*
:Default:
    None -- Maximum value of 60 seconds.

Time to delay before returning a response to a completed DATA command. You can
use this to delay testing or simulate lag.

::

    delay = 30

-----

.. _mode:

mode
----

:Syntax:
    **mode** = *accept | bounce | random*
:Default:
    accept -- valid options are:- accept, bounce, random.

::

    mode = random

-----

.. _max_message_size:

max_message_size
----------------

:Syntax:
    **max_message_size** = *bytes*
:Default:
    512000 Bytes (512 KB)
:Added:
    :ref:`2.0.4`

The maximum message size for a message. This includes headers and helps
mitigate a DoS risk.

::

    max_message_size = 1024000

-----

.. _dynamic_switch:

dynamic_switch
--------------

:Syntax:
    **dynamic_switch** = *true | false*
:Default:
    true -- valid options are:- true, false.
:Added:
    :ref:`2.0.6`

The dynamic switch option allows you to enable or disable parsing of dynamic
switches from email headers -- :ref:`dynamic-switches`

::

    dynamic_switch = false

-----

.. _workers:

workers
-------

:Syntax:
    **workers** = *number*
:Default:
    1
:Added:
    :ref:`2.1.0`

The workers option allows you to define how many worker processes to spawn to
handle incoming mail. The absolute minimum is actually 2. Even by setting the
``workers`` value to 1, a supervisor process will always exist meaning that you
would have 1 worker and a supervisor.

-----


STARTTLS
--------

Currently `asyncio` does not have the code in place to make STARTTLS
possible, the STARTTLS verb returns a ``500 Not implemented`` response
until it's possible to implement. --`https://bugs.python.org/review/23749/
<https://bugs.python.org/review/23749/>`_


Optional features (you should probably use)
===========================================

Blackhole has builtin support for the following features. While these are not
required for the service to run, they do improve it the server in various ways.

uvloop
------

    uvloop is a fast, drop-in replacement of the built-in asyncio event loop.
    uvloop is implemented in Cython and uses libuv under the hood.

Using `uvloop <https://github.com/MagicStack/uvloop>`_ with Blackhole is as
simple as installing `libuv <https://github.com/libuv/libuv>`_ and the
Blackhole extra package.

On Debian/Ubuntu it's as simple as installing via APT and Pip respectively.

.. code:: bash

    apt-get install libuv1 libuv1-dev python-dev
    pip install blackhole[uvloop]

setproctitle
------------

`setproctitle <https://pypi.python.org/pypi/setproctitle>`_ is a simple library
that allows Blackhole to set a more `ps aux`-friendly output for the blackhole
processes.

.. code:: bash

    # without setproctitle
    python3.6 /home/kura/.virtualenvs/blackhole/bin/blackhole -c test.conf -d
    python3.6 /home/kura/.virtualenvs/blackhole/bin/blackhole -c test.conf -d
    # with setproctitle
    blackhole: master
    blackhole: worker

You can install setproctitle with the Blackhole extra package.

.. code:: bash

    pip install blackhole[setproctitle]


Installing the init.d/rc.d scripts
==================================

The init script depends on */etc/blackhole.conf* being in place and configured.

Blackhole comes with a script that works with init.d/rc.d, to install it copy it
from the *init.d/YOUR_DISTRO* folder in the root directory of this project to
*/etc/init.d/*.

The init scripts can be found `here`_.

.. _here: https://github.com/kura/blackhole/tree/master/init.d

i.e. for Debian/Ubuntu users, mv the file from *init.d/debian-ubuntu/* to */etc/init.d/*.

Then make sure it's executable

.. code-block:: bash

  chmod +x /etc/init.d/blackhole

To make blackhole start on a reboot use the following::

  update-rc.d blackhole defaults


CHANGES
=======
Changes
=======

1.8.0
=====

- Removed bin/blackhole and replaced it with a Python entry point
- Cleaned up the source code and modified error checking
- Major refactoring of tests
- Added a requirement of the mock libary to run tests
- Added a wealth of tests for many new methods and some old ones that were
  absent
- Changed get_mailname method to use __builtin__.open instead of
  __builtin__.file

1.7.0
=====

- Added STARTTLS

1.6.4
=====

- Added delay flag
- Fixed daemonisation issue on PyPy
- Added FQDN to HELO/EHLO
- Removed SMTP 251-253 from responses

1.6.0
=====

- Python 3
- Deprecate ssl_ca_certs_dir


blackhole/opts.py
# (The MIT License)
#
# Copyright (c) 2013 Kura
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the 'Software'), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
blackhole.opts - command line options for the blackhole
server.

Also creates a list of available ports for the server to be
run on, based on configuration and responds with the help
menu when requested or invalid options are given.
"""

import getpass

try:
    import ssl
except ImportError:
    ssl = None

from tornado.options import (define, options)
from tornado import process

from blackhole import __pname__

define('host', default="0.0.0.0", metavar="IP", type=str,
       help="IP address to bind go",
       group="Blackhole")
define('port', default=25, metavar="PORT", type=int,
       help="Port to listen for connections on",
       group="Blackhole")
define('daemon', default=False, metavar="DAEMON", type=bool,
       help="Daemonise the program to the background",
       group="Blackhole")
define('pid', default="/tmp/blackhole.pid", metavar="FILE", type=str,
       help="File to write process information to",
       group="Blackhole")
define("conf", metavar="FILE",
       help="Config file to parse and use. Overrides command line args",
       group="Blackhole")
define("user", default=getpass.getuser(),
       metavar="USER", help="User to drop privs to during run time",
       group="Blackhole")
define("group", default=getpass.getuser(),
       metavar="GROUP", help="Group to drop privs to during run time",
       group="Blackhole")
define("log", default="/tmp/blackhole.log",
       metavar="FILE", help="File to write logs to (not very verbose)",
       group="Blackhole")
define("message_size_limit", default=512000, metavar="BYTES", type=int,
       help="""Maximum size of a message in Bytes, returned in EHLO but"""
            """\n%-37snot enforced""" % "",
       group="Blackhole")
define("timeout", default=60, metavar="SECONDS", type=int,
       help="""Number of seconds to wait before dropping a connection due"""
            """to inactivity""", group="Blackhole")

define('workers', default=None, metavar="NUM", type=int,
       help="""Number of worker processes to spawn."""
            """(default: # of CPUs/Cores - 2 + 1 master)""",
       group="Workers")

define('debug', default=False, metavar="BOOL", type=bool,
       help="Enable/disable debug logging mode. Causes a lot of disk I/O",
       group="Debug")

define("switch_mode", default=False, type=bool,
       metavar="HELO_MODE", help="""Change mode based on MAIL FROM/RCPT TO."""
                                 """Please see documentation""",
       group="Mode")
define("mode", default="accept",
       metavar="MODE", help="""Mode to run blackhole in (accept, bounce, """
                            """random,\n%-37sunavailable, offline)""" % "",
       group="Mode")

define('ssl', default=True, metavar="BOOL", type=bool,
       help="Enable/disable SSL",
       group="Blackhole SSL")
define('ssl_port', default=465, metavar="PORT", type=int,
       help="Port to listen for SSL connections on",
       group="Blackhole SSL")
define('ssl_cert', default=None, metavar="PATH", type=str,
       help="SSL Certificate",
       group="Blackhole SSL")
define('ssl_key', default=None, metavar="PATH", type=str,
       help="SSL Private Key",
       group="Blackhole SSL")
define('ssl_ca_certs_dir', default=None, metavar="PATH", type=str,
       help="SSL CA Certificates directory",
       group="Blackhole SSL")

# Eugh, had to replicate to shut up Tornado..
define('version', help="Return program version")
define('v', help="Return program version")


def ports():
    """
    Create and return a list of sockets we need to create.

    A maximum of two will be returned, default is standard
    (std) and SSL (ssl) it is enabled.
    """
    socks_list = ['std']
    # We shouldn't be able to create an SSL socket if it's
    # disabled or if there's non SSL libary
    if options.ssl and ssl:
        socks_list.append('ssl')
    return socks_list


def workers():
    if options.workers is None:
        return process.cpu_count() - 2
    return options.workers


def deprecated_opts():
    dep = ('ssl_ca_certs_dir', )
    for d in dep:
        if getattr(options, d) is not None:
            print("Deprecated option: {}".format(d))


def print_help():
    """Prints all the command line options to stdout."""
    print("Usage: {} [OPTIONS] (start|stop|status)".format(__pname__))
    print("")
    print("  -v, --%-26s %s" % ("version", "Print out program version"))
    print("  -h, --%-26s %s" % ("help", "Show this help information"))
    by_group = {}
    opts = {}
    for option, value in options._options.items():
        # hack to bypass Tornado options
        if option.startswith(("workers", "host", "port", "pid",
                              "conf", "user", "group", "log",
                              "debug", "mode", "ssl", "daemon",
                              "ssl_port", "ssl_cert", "ssl_key",
                              "message_size_limit", "timeout",)):
            if not option.startswith(("log_", "logging", "ssl_ca_certs_dir")):
                opts[option] = value
    for option in opts.values():
        by_group.setdefault(option.group_name, []).append(option)

    for filename, o in sorted(by_group.items()):
        if filename and not filename.endswith("log.py"):
            print("")
            print(filename)
            l = ""
            for _ in range(0, len(filename)):
                l += "-"
            print(l)
            print("")
        o.sort(key=lambda option: option.name)
        for option in o:
            prefix = option.name
            if option.metavar:
                prefix += "=" + option.metavar
            print("  --%-30s %s" % (prefix, option.help or ""))
            if option.name == "switch_mode":
                print("")
                print("""%-34s @accept.blackhole.io""" % "")
                print("""%-34s @bounce.blackhole.io""" % "")
                print("""%-34s @random.blackhole.io""" % "")
                print("""%-34s @unavailable.blackhole.io""" % "")
                print("""%-34s @offline.blackhole.io""" % "")
            if option.name == "mode":
                print("")
                print("""%-34s accept - accept all email with code 250, 251,"""
                      """ 252 or 253""" % "")
                print("""%-34s bounce - bounce all email with a random code,"""
                      """\n%-37sexcluding 250, 251, 252, 253""" % ("", ""))
                print("""%-34s random - randomly accept or bounce all email"""
                      """ with a random code""" % "")
                print("""%-34s unavailable - server always respondes with"""
                      """ code 421\n%-37s- service is unavailable""" %
                      ("", ""))
                print("""%-34s offline - server always responds with code"""
                      """ 521 - server\n%-37sdoes not accept mail""" %
                      ("", ""))
    print("")
